### PR TITLE
Polar Phase Field Model

### LINKED ISSUE #423: Implement polar phase field model
Implement the model by Momeni and Levitas from PHYSICAL REVIEW B 89, 184102 (2014)

> An advanced three-phase phase field approach (PFA) is suggested for a nonequilibrium phase interface that contains an intermediate phase, in particular, a solid-solid interface with a nanometer-sized intermediate melt (IM). A thermodynamic potential in the polar order parameters is developed that satisfies all thermodynamic equilibrium and stability conditions. The special form of the gradient energy allowed us to include the interaction of two solid-melt interfaces via an intermediate melt and obtain a well-posed problem and mesh-independent solutions.

- [x] Document model
- [ ] Assess which kernels from the phase field module can be used
- [x] Implement remaining kernels
- [ ] Add action to simplify model setup

### PR DESCRIPTION
**Pull request by dschwen**
Implementation of the Momeni and Levitas (2014) phase field model.

Closes #423

### CODE REVIEW

**Comment by dschwen** [comment]
The tests that fail on Civet with 
```
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:23: field node 'E_ss' does not hold a float-typed value (val=''6.2415090744607635'')
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:24: field node 'E_ss' does not hold a float-typed value (val=''6.2415090744607635'')
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:26: field node 'Delta_ss' does not hold a float-typed value (val=''${units 3e-9 m -> nm}'')
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:27: field node 'Delta_ss' does not hold a float-typed value (val=''${units 3e-9 m -> nm}'')
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:28: field node 'Delta_md' does not hold a float-typed value (val=''${fparse Delta_ss/KD}'')
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:29: field node 'Delta_mb' does not hold a float-typed value (val=''${fparse Delta_ss/KD}'')
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:43: invalid float syntax for parameter: ICs/Upsilon/beta10=${beta1}
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:44: invalid float syntax for parameter: ICs/Upsilon/beta20=${beta2}
polar_phasefield.ic: /opt/civet/build_0/magpie/tests/polar_phasefield/ic.i:45: invalid float syntax for parameter: ICs/Upsilon/p=${P4}
```

pass locally for me:
```
(moose) [schwd][~/Programs/magpie]> ./run_tests --re polar
polar_phasefield.action ................................................................................... OK
polar_phasefield.ic ....................................................................................... OK
--------------------------------------------------------------------------------------------------------------
Ran 2 tests in 1.1 seconds.
2 passed, 0 skipped, 0 pending, 0 failed
```

Any idea what's going on with that @rwcarlsen ?

**Comment by rwcarlsen** [comment]
What happens if you remove the single quotes from around the brace expressions?

**Comment by dschwen** [comment]
I suppose it works on both platforms then. But those are added by `hit format`...

**Comment by dschwen** [comment]
I have a feeling we already talked about this issue and I thought it got fixed. It was about the Node "kind" not getting set correctly.

File: doc/content/source/actions/PolarPhaseFieldAction.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# PolarPhaseFieldAction
+
+!syntax description /Magpie/PolarPhaseField/PolarPhaseFieldAction
+
+This action simplifies the setup of the [polar phase field model](/PolarPhaseField/index.md)
```
Comment (recuero) [inline doc/content/source/actions/PolarPhaseFieldAction.md:5]: Should this say what `kernel`s this action sets up

File: examples/polar_pfm/action.i [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  xmax = 40
+  ymax = 1
+[]
+
+[Variables]
+  [./Upsilon]
+  [../]
+  [./theta]
+  [../]
+[]
```
Comment (recuero) [inline examples/polar_pfm/action.i:21]: Do we still need to create these variables when using the action?

File: tests/polar_phasefield/tests [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  [./action]
+    type = 'Exodiff'
+    input = 'action.i'
+    exodiff = 'action_out.e'
+  [../]
+  [./ic]
+    type = 'Exodiff'
+    input = 'ic.i'
+    exodiff = 'ic_out.e'
+  [../]
```
Comment (recuero) [inline tests/polar_phasefield/tests:11]: May be handy to add the `no_action` example here to check its results against the `action`'s.

File: doc/content/source/actions/PolarPhaseFieldAction.md (hunk 1 above)
Comment (recuero) [reply doc/content/source/actions/PolarPhaseFieldAction.md:5]: I'd also add a practical application for the use of this kernel/action

File: examples/polar_pfm/verification/run.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+for var in upsilon theta
+do
+  for T in `seq 392 10 532`
```
Comment (recuero) [inline examples/polar_pfm/verification/run.sh:5]: Does this file need to be checked in?

File: include/ics/PolarPFMInterfaceIC.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  virtual Real value(const Point & p) override;
+
+protected:
+  const VariableValue & _theta;
+
+  const Real _a_beta;
+  const Real _beta10;
+  const Real _beta20;
+  const Real _p;
```
Comment (recuero) [inline include/ics/PolarPFMInterfaceIC.h:31]: May be good to leave a Doxygen comment describing that these parameters describe

File: include/materials/PolarPFMBetaS0.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ */
+class PolarPFMBetaS0 : public DerivativeParsedMaterialHelper, public ExpressionBuilder
+{
+public:
+  static InputParameters validParams();
+
+  PolarPFMBetaS0(const InputParameters & parameters);
+
+protected:
+  EBTerm _theta;
```
Comment (recuero) [inline include/materials/PolarPFMBetaS0.h:25]: What's the purpose of using the `ExpressionBuilder` for `_theta`?

File: src/kernels/PolarPFMDerivative.C [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  params.addRequiredParam<MaterialPropertyName>(
+      "F", "Material property of which the kernel variable derivative will be taken of");
+  return params;
+}
+
+PolarPFMDerivative::PolarPFMDerivative(const InputParameters & parameters)
+  : DerivativeMaterialInterface<KernelValue>(parameters),
+    _dpropdu(getMaterialPropertyDerivative<Real>("F", _var.name())),
+    _d2propdu2(getMaterialPropertyDerivative<Real>("F", _var.name(), _var.name()))
+
```
Comment (recuero) [inline src/kernels/PolarPFMDerivative.C:27]: remove space

**Comment by rwcarlsen** [comment]
@dschwen - I vaguely recall dealing with an issue like this as well.  I have no idea how it would be working locally for you and not on CIVET.

File: doc/content/source/actions/PolarPhaseFieldAction.md (hunk 1 above)
Comment (dschwen) [reply doc/content/source/actions/PolarPhaseFieldAction.md:5]: There are verification examples and a test.

File: examples/polar_pfm/verification/run.sh (hunk 4 above)
Comment (dschwen) [reply examples/polar_pfm/verification/run.sh:5]: It helps generating the data for the for the plots in the jupyter notebook. I could also have the notebook run magpie though

Comment (recuero) [reply examples/polar_pfm/verification/run.sh:5]: That's good. Just making sure.

File: doc/content/source/actions/PolarPhaseFieldAction.md (hunk 1 above)
Comment (recuero) [reply doc/content/source/actions/PolarPhaseFieldAction.md:5]: I guess I was referring to elaborating on potentially practical applications. The examples aren't documented, so it doesn't help much.

File: examples/polar_pfm/action.i (hunk 2 above)
Comment (dschwen) [reply examples/polar_pfm/action.i:21]: yeah, the action does not add those

File: doc/content/source/actions/PolarPhaseFieldAction.md (hunk 1 above)
Comment (dschwen) [reply doc/content/source/actions/PolarPhaseFieldAction.md:5]: good point

File: examples/polar_pfm/verification/run.sh (hunk 4 above)
Comment (dschwen) [reply examples/polar_pfm/verification/run.sh:5]: I reworked the jupyter notebook to run the verification cases and generate all movies and plots

File: include/materials/PolarPFMBetaS0.h (hunk 6 above)
Comment (dschwen) [reply include/materials/PolarPFMBetaS0.h:25]: It is holding a symbolic reference to the theta variable. ExpressionBuilder is used to build the symbolic derivatives of the expression this material is representing.

File: doc/content/source/actions/PolarPhaseFieldAction.md (hunk 1 above)
Comment (dschwen) [reply doc/content/source/actions/PolarPhaseFieldAction.md:5]: I added an explanation for the physical meaning of the model. For more details I would refer the reader to the original publication, which is cited.

File: tests/polar_phasefield/tests (hunk 3 above)
Comment (dschwen) [reply tests/polar_phasefield/tests:16]: added

**Review by recuero — APPROVED** [review]

**Pull request merged**